nexus/db-queries/src/db/datastore/silo_auth_settings.rs

```diff
@@ -0,0 +1,114 @@
+use super::DataStore;
+use crate::authz;
+use crate::context::OpContext;
+use async_bb8_diesel::AsyncRunQueryDsl;
+use diesel::prelude::*;
+use nexus_db_errors::ErrorHandler;
+use nexus_db_errors::public_error_from_diesel;
+use nexus_db_lookup::DbConnection;
+use nexus_db_model::SiloAuthSettings;
+use nexus_db_model::SiloAuthSettingsUpdate;
+use omicron_common::api::external::DeleteResult;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::ResourceType;
+use omicron_common::api::external::UpdateResult;
+
+// Directly modeled on settings query functions.
+
+impl DataStore {
+    /// Creates new settings for a silo. This is grouped with silo creation
+    /// and shouldn't be called outside of that flow.
+    ///
+    /// An authz check _cannot_ be performed here because the authz initialization
+    /// isn't complete and will lead to a db deadlock.
+    ///
+    /// See <https://github.com/oxidecomputer/omicron/blob/07eb7dafc20e35e44edf429fcbb759cbb33edd5f/nexus/db-queries/src/db/datastore/rack.rs#L407-L410>
+    pub async fn silo_auth_settings_create(
+        &self,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        authz_silo: &authz::Silo,
+        settings: SiloAuthSettings,
+    ) -> Result<(), Error> {
+        let silo_id = authz_silo.id();
+        use nexus_db_schema::schema::silo_auth_settings;
+
+        diesel::insert_into(silo_auth_settings::table)
+            .values(settings)
+            .execute_async(conn)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel(
+                    e,
+                    ErrorHandler::Conflict(
+                        ResourceType::SiloAuthSettings,
+                        &silo_id.to_string(),
+                    ),
+                )
+            })
+            .map(|_| ())
+    }
+
+    pub async fn silo_auth_settings_delete(
+        &self,
+        opctx: &OpContext,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        authz_silo: &authz::Silo,
+    ) -> DeleteResult {
+        // Given that the settings right now are somewhat of an extension of the
+        // Silo we just check for delete permission on the silo itself.
+        opctx.authorize(authz::Action::Delete, authz_silo).await?;
```

Comment:
hm, so this only happens as part of deleting the whole silo, right? maybe instead of doing the authz check here, this should just be a `pub(crate)` function instead. that way, it's internal to `nexus_db_queries` and is just only ever called when deleting a silo.
```suggestion
    pub(crate) async fn silo_auth_settings_delete(
        &self,
        opctx: &OpContext,
        conn: &async_bb8_diesel::Connection<DbConnection>,
        authz_silo: &authz::Silo,
    ) -> DeleteResult {
```